<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo"  />
    </a>
</p>


# **Launch Sites Locations Analysis with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [1]:
import piplite
await piplite.install(['folium'])

await piplite.install(['pandas'])

In [2]:
import folium
import pandas as pd

In [3]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module\_3/DV0101EN-3-5-1-Generating-Maps-in-Python-py-v2.0.ipynb)


## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [4]:
# Download and read the `spacex_launch_geo.csv`
from js import fetch
import io

URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv'
resp = await fetch(URL)
spacex_csv_file = io.BytesIO((await resp.arrayBuffer()).to_py())
spacex_df=pd.read_csv(spacex_csv_file)

Now, you can take a look at what are the coordinates for each site.


In [5]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [6]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [7]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [8]:
# Initial the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)
# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. In addition, add Launch site name as a popup label
site_name = [name for name in launch_sites_df['Launch Site']]
site_name

['CCAFS LC-40', 'CCAFS SLC-40', 'KSC LC-39A', 'VAFB SLC-4E']

In [9]:
launch_sites_df.loc[launch_sites_df['Launch Site'] == 'CCAFS LC-40']['Lat']

0    28.562302
Name: Lat, dtype: float64

In [10]:
for site in site_name:
    circle = folium.Circle([launch_sites_df.loc[launch_sites_df['Launch Site'] == site]['Lat'],launch_sites_df.loc[launch_sites_df['Launch Site'] == site]['Long']], 
                           radius=1000, color='#d35400', 
                           fill=True).add_child(folium.Popup(site))
    
    # Create a blue circle at [site] coordinate with a icon showing its name
    marker = folium.map.Marker(
        [launch_sites_df.loc[launch_sites_df['Launch Site'] == site]['Lat'],launch_sites_df.loc[launch_sites_df['Launch Site'] == site]['Long']],
        
        # Create an icon as a text label
        icon=DivIcon(
            icon_size=(20,20),
            icon_anchor=(0,0),
            html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % site,
            )
        )
    site_map.add_child(circle)
    site_map.add_child(marker)

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>


In [11]:
site_map

Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?
*   Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


# Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [12]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [14]:
marker_cluster = MarkerCluster()

*TODO:* Create a new column in `launch_sites` dataframe called `marker_color` to store the marker colors based on the `class` value


In [22]:
# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red
spacex_df['marker_color'] = spacex_df['class']
spacex_df

Launch Site        Lat        Long  class  marker_color
0    CCAFS LC-40  28.562302  -80.577356      0             0
1    CCAFS LC-40  28.562302  -80.577356      0             0
2    CCAFS LC-40  28.562302  -80.577356      0             0
3    CCAFS LC-40  28.562302  -80.577356      0             0
4    CCAFS LC-40  28.562302  -80.577356      0             0
5    CCAFS LC-40  28.562302  -80.577356      0             0
6    CCAFS LC-40  28.562302  -80.577356      0             0
7    CCAFS LC-40  28.562302  -80.577356      0             0
8    CCAFS LC-40  28.562302  -80.577356      0             0
9    CCAFS LC-40  28.562302  -80.577356      0             0
10   CCAFS LC-40  28.562302  -80.577356      0             0
11   CCAFS LC-40  28.562302  -80.577356      0             0
12   CCAFS LC-40  28.562302  -80.577356      0             0
13   CCAFS LC-40  28.562302  -80.577356      0             0
14   CCAFS LC-40  28.562302  -80.577356      0             0
15   CCAFS LC-40  28.562302  -80.577356      0             0
16   CCAFS LC-40  28.562302  -80.577356      0             0
17   CCAFS LC-40  28.562302  -80.577356      1             1
18   CCAFS LC-40  28.562302  -80.577356      1             1
19   CCAFS LC-40  28.562302  -80.577356      0             0
20   CCAFS LC-40  28.562302  -80.577356      1             1
21   CCAFS LC-40  28.562302  -80.577356      1             1
22   CCAFS LC-40  28.562302  -80.577356      1             1
23   CCAFS LC-40  28.562302  -80.577356      0             0
24   CCAFS LC-40  28.562302  -80.577356      1             1
25   CCAFS LC-40  28.562302  -80.577356      1             1
26   VAFB SLC-4E  34.632834 -120.610745      0             0
27   VAFB SLC-4E  34.632834 -120.610745      0             0
28   VAFB SLC-4E  34.632834 -120.610745      1             1
29   VAFB SLC-4E  34.632834 -120.610745      1             1
30   VAFB SLC-4E  34.632834 -120.610745      1             1
31   VAFB SLC-4E  34.632834 -120.610745      1             1
32   VAFB SLC-4E  34.632834 -120.610745      0             0
33   VAFB SLC-4E  34.632834 -120.610745      0             0
34   VAFB SLC-4E  34.632834 -120.610745      0             0
35   VAFB SLC-4E  34.632834 -120.610745      0             0
36    KSC LC-39A  28.573255  -80.646895      1             1
37    KSC LC-39A  28.573255  -80.646895      0             0
38    KSC LC-39A  28.573255  -80.646895      1             1
39    KSC LC-39A  28.573255  -80.646895      1             1
40    KSC LC-39A  28.573255  -80.646895      0             0
41    KSC LC-39A  28.573255  -80.646895      1             1
42    KSC LC-39A  28.573255  -80.646895      1             1
43    KSC LC-39A  28.573255  -80.646895      0             0
44    KSC LC-39A  28.573255  -80.646895      1             1
45    KSC LC-39A  28.573255  -80.646895      1             1
46    KSC LC-39A  28.573255  -80.646895      1             1
47    KSC LC-39A  28.573255  -80.646895      1             1
48    KSC LC-39A  28.573255  -80.646895      1             1
49  CCAFS SLC-40  28.563197  -80.576820      1             1
50  CCAFS SLC-40  28.563197  -80.576820      1             1
51  CCAFS SLC-40  28.563197  -80.576820      0             0
52  CCAFS SLC-40  28.563197  -80.576820      0             0
53  CCAFS SLC-40  28.563197  -80.576820      0             0
54  CCAFS SLC-40  28.563197  -80.576820      1             1
55  CCAFS SLC-40  28.563197  -80.576820      0             0

In [26]:
spacex_df['marker_color'] = spacex_df['marker_color'].apply(lambda x: 'red' if x == 0 else 'green')
spacex_df['marker_color']

0       red
1       red
2       red
3       red
4       red
5       red
6       red
7       red
8       red
9       red
10      red
11      red
12      red
13      red
14      red
15      red
16      red
17    green
18    green
19      red
20    green
21    green
22    green
23      red
24    green
25    green
26      red
27      red
28    green
29    green
30    green
31    green
32      red
33      red
34      red
35      red
36    green
37      red
38    green
39    green
40      red
41    green
42    green
43      red
44    green
45    green
46    green
47    green
48    green
49    green
50    green
51      red
52      red
53      red
54    green
55      red
Name: marker_color, dtype: object

*TODO:* For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [31]:
# Add marker_cluster to current site_map
site_map.add_child(marker_cluster)

# for each row in spacex_df data frame
# create a Marker object with its coordinate
# and customize the Marker's icon property to indicate if this launch was successed or failed, 
# e.g., icon=folium.Icon(color='white', icon_color=row['marker_color']
for index, record in spacex_df.iterrows():
    # TODO: Create and add a Marker cluster to the site map
    marker = folium.Marker(location=[record[1],record[2]],
                          icon=folium.Icon(color='white', icon_color=record[4]))
    marker_cluster.add_child(marker)

site_map

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png">
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png">
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


# TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [32]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


In [33]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

*TODO:* Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [47]:
coast_lat = 28.56336
coast_lon = -80.56796
launch_site_lat = spacex_df['Lat'][0]
launch_site_lon = spacex_df['Long'][0]
# find coordinate of the closet coastline
# e.g.,: Lat: 28.56367  Lon: -80.57163
distance_airport = calculate_distance(launch_site_lat, launch_site_lon, coast_lat, coast_lon)
distance_airport

0.925477797003777

In [48]:
coordinate = [coast_lat,coast_lon]

In [49]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
distance_marker = folium.Marker(coordinate,
                                icon=DivIcon(icon_size=(20,20),
                                             icon_anchor=(0,0),
                                             html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_airport)))

In [51]:
site_map.add_child(distance_marker)

*TODO:* Draw a `PolyLine` between a launch site to the selected coastline point


In [50]:
coordinates = [coordinate,[launch_site_lat,launch_site_lon]]
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
lines=folium.PolyLine(locations=coordinates,popup=distance_airport, weight=2)
site_map.add_child(lines)

Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png">
</center>


*TODO:* Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


In [ ]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site


In [52]:
airport_lat = 28.5091
airport_lon = -80.79617
# find coordinate of the closest airport
coordinate2 = [airport_lat,airport_lon]
distance_airport2 = calculate_distance(launch_site_lat, launch_site_lon, airport_lat, airport_lon)
distance_marker = folium.Marker(coordinate2,
                                icon=DivIcon(icon_size=(20,20),
                                             icon_anchor=(0,0),
                                             html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_airport2)))

In [65]:
# City
city_lat = 28.38672
city_lon = -80.60034
# find coordinate of the closest airport
coordinate3 = [city_lat,city_lon]
distance_airport3 = calculate_distance(launch_site_lat, launch_site_lon, city_lat, city_lon)
distance_marker = folium.Marker(coordinate3,
                                icon=DivIcon(icon_size=(20,20),
                                             icon_anchor=(0,0),
                                             html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_airport3)))
site_map.add_child(distance_marker)
coordinates = [[city_lat,city_lon],[launch_site_lat,launch_site_lon]]
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
lines=folium.PolyLine(locations=coordinates,popup=distance_airport3, weight=2)
site_map.add_child(lines)

In [66]:
# Railway :
railway_lat = 28.57171
railway_lon = -80.58527
# find coordinate of the closest airport
coordinate4 = [railway_lat,railway_lon]
distance_airport4 = calculate_distance(launch_site_lat, launch_site_lon, railway_lat, railway_lon)
distance_marker = folium.Marker(coordinate4,
                                icon=DivIcon(icon_size=(20,20),
                                             icon_anchor=(0,0),
                                             html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_airport4)))
site_map.add_child(distance_marker)
coordinates = [[railway_lat,railway_lon],[launch_site_lat,launch_site_lon]]
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
lines=folium.PolyLine(locations=coordinates,popup=distance_airport4, weight=2)
site_map.add_child(lines)

In [67]:
# Highway : 
highway_lat = 28.56234
highway_lon = -80.57061
# find coordinate of the closest airport
coordinate5 = [highway_lat,highway_lon]
distance_airport5 = calculate_distance(launch_site_lat, launch_site_lon, highway_lat, highway_lon)
distance_marker = folium.Marker(coordinate5,
                                icon=DivIcon(icon_size=(20,20),
                                             icon_anchor=(0,0),
                                             html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_airport5)))
site_map.add_child(distance_marker)
coordinates = [[highway_lat,highway_lon],[launch_site_lat,launch_site_lon]]
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
lines=folium.PolyLine(locations=coordinates,popup=distance_airport5, weight=2)
site_map.add_child(lines)

After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways?
*   Are launch sites in close proximity to highways?
*   Are launch sites in close proximity to coastline?
*   Do launch sites keep certain distance away from cities?

Also please try to explain your findings.


Answers to each question: 
- Are launch sites in close proximity to railways?
  Yes for instance for launch_site CCAFS LC-40 where the closest railway is at a distance of 1.30 km
- Are launch sites in close proximity to highways?
  Yes for instance for launch_site CCAFS LC-40 where the closest railway is at a distance of 0.66 km
- Are launch sites in close proximity to coastline?
  Yes for instance for launch_site CCAFS LC-40 where the closest railway is at a distance of 0.93 km
- Do launch sites keep certain distance away from cities?
  In the case of launch_site CCAFS LC-40 the closest city is Cape Canaveral and the distance is 19.66 km which I guess should be far enough from the launch and its possible consequences.

# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Pratiksha Verma](https://www.linkedin.com/in/pratiksha-verma-6487561b1/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork865-2022-01-01)


## Change Log


| Date (YYYY-MM-DD) | Version | Changed By      | Change Description      |
| ----------------- | ------- | -------------   | ----------------------- |
| 2022-11-09        | 1.0     | Pratiksha Verma | Converted initial version to Jupyterlite|


### <h3 align="center"> IBM Corporation 2022. All rights reserved. <h3/>
